### DISCUSSION TITLE
What are all the steps needed to enable releasing on chocolatey?

### DISCUSSION THREAD

**Discussion by wimdeblauwe** [jreleaser/jreleaser#1382] [Q&A]
I have a working JReleaser setup that already releases to homebrew. I would like to add Chocolatey support for Windows users. I had a look to https://github.com/jreleaser/jreleaser/blob/main/jreleaser.yml#LL291C5-L291C15 for inspiration and did commit https://github.com/wimdeblauwe/ttcli/commit/ee9602d585f4d3cc2c7cea7dee4ebf9b603d48d0 to have those settings defined in my Maven `pom.xml`. 
After I ran a release build, a new repo was created https://github.com/wimdeblauwe/chocolatey-bucket but there is nothing in there. There was no error during the build, but I assume the empty repo is probably not the intention.

Can somebody point me to what I need to do? Do I need to do something manually in that empty repo? Do I need to add other properties in my Maven `pom.xml`? Is it possible to use `chocolatey-ttcli` as the repo name?

**Comment by wimdeblauwe** [comment]
The `trace.log` does not indicate any error, yet nothing is actually pushed on the `chocolatey-bucket` git repo:

```
[INFO]      [chocolatey] publishing ttcli distribution
[DEBUG]     [chocolatey] creating props for ttcli/chocolatey
[DEBUG]     [chocolatey] filling distribution properties into props
[DEBUG]     [chocolatey] filling git properties into props
[DEBUG]     [chocolatey] filling artifact properties into props
[DEBUG]     [chocolatey] filling packager properties into props
[INFO]      [chocolatey] setting up repository wimdeblauwe/chocolatey-bucket
[DEBUG]     [chocolatey] locating repository wimdeblauwe/chocolatey-bucket
[DEBUG]     [chocolatey] lookup repository wimdeblauwe/chocolatey-bucket
[DEBUG]     [chocolatey] lookup repository wimdeblauwe/chocolatey-bucket
[DEBUG]     [chocolatey] cloning https://github.com/wimdeblauwe/chocolatey-bucket.git
[DEBUG]     [chocolatey] copying files from target/jreleaser/package/ttcli/chocolatey
[DEBUG]     [chocolatey] copying files from /home/runner/work/ttcli/ttcli/target/jreleaser/package/ttcli/chocolatey to /tmp/jreleaser-chocolatey-bucket14570349979511427769
[DEBUG]     [chocolatey] setting up commit
[DEBUG]     [chocolatey] tagging local repository with 0.9.2
[INFO]      [chocolatey] pushing to wimdeblauwe/chocolatey-bucket
[DEBUG]     [chocolatey] pushing commit to remote
```

**Reply by aalmiray** [reply]
Yes, I believe the repository name has no constraints. Could you post the resolved model found in `trace.log`?

**Reply by wimdeblauwe** [reply]
Not sure what you mean, but you can see the full trace.log file at https://github.com/wimdeblauwe/ttcli/actions/runs/4925363153 under "jreleaser-release-output"

**Reply by wimdeblauwe** [reply]
I have now added an extra secret `JRELEASER_CHOCOLATEY_GITHUB_TOKEN` (With workflow scope) to the repo and also used that secret in my release.yml workflow file. As a result, the https://github.com/wimdeblauwe/chocolatey-bucket repo now has some files added. However, the push.yml action on that repo fails, stating that the API key is missing. I do have the `JRELEASER_CHOCOLATEY_API_KEY` env variable set in my original repo. Do I need to do something in this special chocolatey-bucket repo as well manually?

**Reply by aalmiray** [reply]
Ah yea, the default `GITHUB_TOKEN` is not enough when pushing to a different repository. The docs should mention this fact. There should had been a failure registered in the log because of insufficient write permissions. That's a bug.

**Reply by aalmiray** [reply]
The bucket repository expects a secret named `CHOCOLATEY_API_KEY`.

**Reply by wimdeblauwe** [reply]
Yes, I added that secret now to the bucket repository and the push.yml workflow now succeeded. Thanks!

Do you want me to open a bug for the failure that is not showing?

**Reply by aalmiray** [reply]
Yes, please. We used to log an error in the past. This means likely we changed something in this regard without checking for this use case.

**Reply by wimdeblauwe** [reply]
I created https://github.com/jreleaser/jreleaser/issues/1383

**Comment by wimdeblauwe** [comment]
To fully answer my own question, these are the steps I needed to do:

1. Add this to the `BINARY` distribution section of the jreleaser maven plugin configuration in your `pom.xml`:
```
                                      <chocolatey>
                                            <active>ALWAYS</active>
                                            <remoteBuild>true</remoteBuild>
                                            <title>ttcli</title>
                                            <bucket>
                                                <active>RELEASE</active>
                                            </bucket>
                                        </chocolatey>
```                                     
See https://github.com/wimdeblauwe/ttcli/blob/0.9.2/pom.xml#L236-L243

2. In your `release.yml` workflow for GitHub Actions, add 2 env variables:
```
JRELEASER_CHOCOLATEY_GITHUB_TOKEN: ${{ secrets.JRELEASER_CHOCOLATEY_GITHUB_TOKEN }}
JRELEASER_CHOCOLATEY_API_KEY: ${{ secrets.CHOCOLATEY_API_KEY }}
```
See https://github.com/wimdeblauwe/ttcli/blob/0.9.2/.github/workflows/release.yml#L142-L143  

The `JRELEASER_CHOCOLATEY_GITHUB_TOKEN` should be created via https://github.com/settings/tokens. There, create a "Personal Access Token (classic)" with "workflow" and "admin:org" permissions. After creating this, add it to your repository via "Settings" tab, selecting "Secrets and variables" > "Actions" and use the "New repository secret" button.

For the CHOCOLATEY_API_KEY, you need to create an account with Chocolatey at https://community.chocolatey.org/account. After registration, you can get an API key there. Add it also as a repository secret in GitHub.

After the first run, a new repository will be created called `chocolatey-bucket`. There will be a push.yml workflow in there that will fail because of the missing CHOCOLATEY_API_KEY also there. Add the API key as a repository secret in that new repository and re-run the action that failed. The publishing to Chocolatey should be working now.
